### PR TITLE
Set agent to 'started' upon completing reconnect to New Relic servers

### PR DESCRIPTION
**Pull request by michaelgoin**
## Proposed Release Notes

* Fixed bug where `API.shutdown` would not harvest or keep process active effectively after an agent restart.

  The agent will now correctly update its state to 'started' after a reconnect has completed.

## Links

* Fixes: https://github.com/newrelic/node-newrelic/issues/789

## Details

Did a minor refactor to consolidate the agent's handling of connection to try to avoid similar issues in the future.

Flagging as 'medium' risk because this is a code path that can impact all data. I have manually tested on a few flavors of application as well as added additional test coverage but worth extra care in review.

### CODE REVIEW

File: lib/agent.js
Comment (michaelgoin) [inline lib/agent.js:462]: Not obvious from refactor but I introduced an async hop via `setImmediate` for the case where aggregators are simply started for consistency with the "immediate" harvest case that gets scheduled via `setTimeout`.

File: lib/agent.js
Comment (michaelgoin) [inline lib/agent.js:447]: onConnect now handles consistency of setting up aggregators and starting them, and setting agent state to 'started.'

Comment (michaelgoin) [reply lib/agent.js:447]: Given there's now always an async hop, moving `setState` to after the harvest scheduling should have no impact. But mentioning since that was a change. I don't see anything in the starting of the aggregators, etc. that relies on that already being set.

**Review by bizob2828 — COMMENTED** [review]
looks great overall, just a few questions on the testing front

File: lib/collector/api.js
Comment (michaelgoin) [reply lib/collector/api.js:521]: yeah, sometimes i like to do that sometimes i don't. it can be harder to debug.

File: test/integration/api/shutdown.tap.js
Comment (michaelgoin) [reply test/integration/api/shutdown.tap.js:62]: The issue being fixed is explicitly for the case where `collectPendingData` is true. Otherwise, the agent state of 'started' doesn't matter.

I'm all for additional test coverage here, just wasn't the focus of this PR.

File: lib/collector/api.js
Comment (michaelgoin) [reply lib/collector/api.js:521]: i'll go ahead and make this change real quick

File: test/integration/api/shutdown.tap.js
Comment (bizob2828) [reply test/integration/api/shutdown.tap.js:62]: yea that was my only thought, just while we're in here adding some missing shutdown scenarios. I'll defer to you deciding if you want to add this test

File: test/integration/api/shutdown.tap.js
Comment (bizob2828) [reply test/integration/api/shutdown.tap.js:54]: might be good to include a little blurb in the test for posterity

File: test/integration/api/shutdown.tap.js
Comment (michaelgoin) [reply test/integration/api/shutdown.tap.js:62]: Think I'm going to leave things for this PR.

This particular case I think the unit test coverage is likely enough, we probably don't need to verify the agent stop function is doing what it should. Covering the timer case outside of a unit test or the errored may add some benefit. I'll look into that separate from this PR so we aren't holding up a fix.

File: test/integration/api/shutdown.tap.js
Comment (michaelgoin) [reply test/integration/api/shutdown.tap.js:54]: sounds good

File: test/integration/api/shutdown.tap.js
Comment (bizob2828) [reply test/integration/api/shutdown.tap.js:62]: works for me

**Review by bizob2828 — APPROVED** [review]

File: test/integration/api/shutdown.tap.js
Comment (bizob2828) [reply test/integration/api/shutdown.tap.js:65]: thanks for updating this. it'll help future team understand this test

**Pull request merged**
